samza-kafka/src/test/scala/org/apache/samza/checkpoint/kafka/TestKafkaCheckpointManager.scala

```diff
@@ -82,7 +83,6 @@ class TestKafkaCheckpointManager extends KafkaServerTestHarness {
     val spec = new KafkaStreamSpec("id", checkpointTopic, checkpointSystemName, 1, 1, props)
     val checkPointManager = Mockito.spy(new KafkaCheckpointManager(spec, new MockSystemFactory, false, config, new NoOpMetricsRegistry))
```

Comment:
Do you know why we need to spy the class under test here rather than mock the underlying objects used by the method that is stubbed?

Reply:
It's not necessary to use the spy, but this test seems to be using some "real" method implementations for some underlying objects, so I guess it is a little awkward to mix in mocking too. I'm not a fan of how this unit test class was written overall (e.g. I think it uses an embedded Kafka/Zk just to run a unit test), but I also didn't have time to refactor everything, so I went with the easiest fix.

Reply:
Fair enough. Better than the test being disabled.

I'd be more worried if all invocations of `KafkaCheckpointManager.getSystemProducer` were stubbed using the spy, but it looks to be specifically targeting the second invocation (the first is during construction), so the test is at least not completely masking away the contract with `SystemFactory`.

You can ignore this comment.